src/graphnet/data/extractors/icecube/i3extractor.py

```diff
@@ -194,27 +209,71 @@ def get_primaries(
                     )
                 ]
 
-            if len(primaries) == 0:
-                primaries = [
-                    p
-                    for p in frame[self.mctree]
-                    if (
-                        p.is_neutrino
-                        & (
-                            p.location_type
-                            == dataclasses.I3Particle.LocationType.InIce.real
-                        )
+                # Sometimes the primary neutrino interacts outside
+                # of the ice. Therefore only further down in the subtree
+                # of the original primary neutrino can you find in-ice
+                # particles
+                if len(primaries) == 0:
+
+                    # get the original primary neutrino(s)
+                    primary_nus = [
+                        p
+                        for p in frame[self.mctree].get_primaries()
+                        if p.is_neutrino
+                    ]
+
+                    # recursively search for in-ice neutrino daughters
+                    primaries = self.find_in_ice_daughters(
+                        frame,
+                        primary_nus,
+                        self.mctree,
                     )
-                ]
+
+                    # This is not expected to happen
```

Comment:
Is this still true, do all NuGen events have a InIce neutrino generated by thedaughter particles?